File: leetcode.com/0725_split_linked_list_in_parts/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def splitListToParts(self, root: ListNode, k: int) -> List[ListNode]:
        l = 0
        node = root

        while node:
            node = node.next
            l += 1

        b = 0   # big block size
        bc = 0  # big block count

        if k > l:
            a = 1
            ac = k
        elif l % k == 0:
            a = l / k
            ac = k
        else:
            a = math.floor(l / k)
            b = math.ceil(l / k)

            if b * (k-1) + a == l:
                bc = k - 1
                ac = 1
            else:
                ac = k - 1
                bc = 1

                while b * bc + a * ac != l:
                    bc += 1
                    ac -= 1

        # construct
        splited = []
        node = root

        # print(b, bc, a, ac)

        for block_size, blocks_count in ((b, bc), (a, ac)):
            for _ in range(blocks_count):
                curr_size = 0
                sln = None
                slr = None

                while node and curr_size < block_size:
                    if not slr:
                        slr = ListNode(node.val)
                        sln = slr
                    else:
                        sln.next = ListNode(node.val)
                        sln = sln.next

                    curr_size += 1
                    node = node.next

                splited.append(slr)

        return splited

File: leetcode.com/0725_split_linked_list_in_parts/solution.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
struct ListNode** splitListToParts(struct ListNode* root, int k, int* returnSize)
{
    int length = 0;

    for (struct ListNode *node = root ; node ; length++)
        node = node->next;

    int chunk_size = length / k;
    int chunk_count = length % k;

    printf("length = %d ; %d, %d\n", length, chunk_size, chunk_count);

    struct ListNode *node = root;
    struct ListNode** splitted = calloc(sizeof(struct ListNode*), k);

    for (int i = 0 ; i < k ; i++)
    {
        struct ListNode *sl = NULL;
        struct ListNode *prev = NULL;

        for (int j = 0 ; node && j < chunk_size + (chunk_count > 0) ? 1 : 0 ; j++, chunk_count--)
        {
            struct ListNode *e = malloc(sizeof(struct ListNode));
            e->val = node->val;
            e->next = NULL;

            if (sl == NULL)
                sl = e;

            if (prev != NULL)
                prev->next = e;

            prev = e;
            node = node->next;
        }

        splitted[i] = sl;
    }

    *returnSize = k;
    return splitted;
}
